kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,389 @@
+    analysis-type, encoded as specified in the `InitRequest`. For Kythe index
+    analyses, this would be `kythe.common.Entry` objects. Other analyzers might
+    emit a "findings" structure.
+
+    Note that contents need not be literally copied. An overlay filesystem,
+    hardlinks, symlinks, or other may be used.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":2,"method":"analyze","params":{"types":["kythe/indexer/go"]}}
+        « {"jsonrpc":"2.0","id":2","result":{"input":"path/to/main.go", "output":"path/to/output.json"}}
+
+-   `vname(VNameRequest)`
+
+    Requests the complete VName to emit for the specified path and signature.
+
+-   `log(LogRequest)`
+
+    A notification from the analyzer to the environment, containing a message
+    string to be logged in association with the current analysis. The
+    environment does not reply.
+
+    Example:
+
+        » {"jsonrpc":"2.0","method":"log","params":{"message":{...}}}
+
+-   `done(LogRequest)`
+
+    A notification from the analyzer to the environment, signifying that the
+    analyzer has completed work on the specified analysis ID. The environment
+    does not reply. After receiving this notification, the driver MUST discard
+    any further requests for the analysis ID (and return an error, if
+    appropriate) unless and until it issues a new analysis reusing that ID.
+
+    Example:
+
+        » {"jsonrpc":"2.0","method":"done","params":{"message":"success"}}
+
+### Errors
+
+In addition to the standard errors defined by JSON-RPC, this protocol defines
+these errors:
+
+-   `ProtocolError` (code: -1). Indicates a violation of the protocol.
+
+## Input filesystem
+
+This RFC essentially postulates that the driver presents the analyzer
+a *limited* virtualization of a filesystem, specific to each
+analysis. In the context of a single analysis:
+
+-   For each file path occurring in the compilation record, the driver MUST 
+    support `open()`, `read()` and `lseek()` (or equivalent) operations. The
+    contents backing such operations MUST be precisely the contents specified
+    in the compilation record. Other file operations MAY be supported.
+
+-   For each directory implied by the file paths in the compilation record, the
+    driver MAY support a subset of read-only directory operations, e.g. in order
+    to provide directory listings via `readdir(3)` or equivalent.
+
+-   For all other paths, the driver MAY fall back to the underlying host
+    filesystem.
+
+## Output
+
+The output file MUST be treated as append-only by the analyzer, with no seek,
+read or other operations permitted. Data MUST written to the output file based
+on the encoding negotiated in the Init request/response.
+
+-   For "json" encoding, the analyzer MUST write the output as
+    [varint](https://developers.google.com/protocol-buffers/docs/encoding#varints) length-prefixed records.
+
+-   For "protobuf" encoding, the analyzer MUST write the output as [delimited
```

Comment:
This is under-specified. What is the format of the message size delimiter?

Reply:
Whatever [standard proto delimited methods](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/MessageLite#writeDelimitedTo-java.io.OutputStream-) do. Is there a canonical place that's documented?

Reply:
There doesn't seem to be a canonical reference for this and there doesn't appear to be an equivalent of writeDelimitedTo() in every language (even though it's the canonical representation of an embedded message).

Reply:
I specified varint, without any canonical reference.